lib/IRGen/LoadableByAddress.cpp

```diff
@@ -0,0 +1,1538 @@
+      pass.storeInstsToMod.push_back(insToInsert);
+      break;
+    }
+    case ValueKind::DebugValueInst: {
+      DebugValueInst *insToInsert = dyn_cast<DebugValueInst>(userIns);
+      assert(insToInsert && "Unexpected cast failure");
+      pass.debugInstsToMod.push_back(insToInsert);
+      break;
+    }
+    case ValueKind::StructExtractInst: {
+      StructExtractInst *instToInsert = dyn_cast<StructExtractInst>(userIns);
+      assert(replaceProjection && "Unexpected cast failure");
+      if (!replaceProjection) {
+        llvm_unreachable("Unexpected usage: Can't replace StructExtractInst");
+      }
+      pass.structExtractInstsToMod.push_back(instToInsert);
+      break;
+    }
+    case ValueKind::SwitchEnumInst: {
+      SwitchEnumInst *instToInsert = dyn_cast<SwitchEnumInst>(userIns);
+      assert(replaceProjection && "Unexpected cast failure");
+      if (!replaceProjection) {
+        llvm_unreachable("Unexpected usage: Can't replace SwitchEnumInst");
+      }
+      pass.switchEnumInstsToMod.push_back(instToInsert);
+      break;
+    }
+    default:
+      llvm_unreachable("Unexpected instruction");
+    }
+  }
+
+  optimizableLoad->replaceAllUsesWith(allocInstr);
+  optimizableLoad->getParent()->erase(optimizableLoad);
+}
+
+static bool usesContainApplies(LoadInst *unoptimizableLoad,
+                               irgen::IRGenModule &Mod) {
+  for (auto *user : unoptimizableLoad->getUses()) {
+    SILInstruction *userIns = user->getUser();
+    switch (userIns->getKind()) {
+    case ValueKind::ApplyInst:
+    case ValueKind::TryApplyInst:
+    case ValueKind::PartialApplyInst: {
+      ApplySite site(userIns);
+      SILValue callee = site.getCallee();
+      if (callee == unoptimizableLoad) {
+        break;
+      }
+      SILType currType = unoptimizableLoad->getType().getObjectType();
+      SILType newSILType = getNewSILType(currType, Mod);
+      if (currType == newSILType) {
+        break;
+      }
+      return true;
+    }
+    default:
+      break;
+    }
+  }
+  return false;
+}
+
+void LoadableStorageAllocation::replaceLoadWithCopyAddrForApplies(
+    LoadInst *unoptimizableLoad) {
+  if (!usesContainApplies(unoptimizableLoad, pass.Mod)) {
+    return;
+  }
+  SILValue value = unoptimizableLoad->getOperand();
+
+  SILBuilder allocBuilder(pass.F->begin()->begin());
+  AllocStackInst *allocInstr =
+      allocBuilder.createAllocStack(getLocForValue(value), value->getType());
+
+  SILBuilder outlinedBuilder(unoptimizableLoad);
+  createOutlinedCopyCall(outlinedBuilder, value, allocInstr, pass);
+
+  // Insert stack deallocations.
+  for (TermInst *termInst : pass.returnInsts) {
+    SILBuilder deallocBuilder(termInst);
+    deallocBuilder.createDeallocStack(allocInstr->getLoc(), allocInstr);
+  }
+
+  SmallVector<Operand *, 8> usersToMod;
+  for (auto *user : unoptimizableLoad->getUses()) {
+    SILInstruction *userIns = user->getUser();
+    switch (userIns->getKind()) {
```

Comment:
As we discussed in person , you can also handle struct_extracts etc, here so that you don't explode using a load (like you do in the plain replaceLoadWithCopyAddr() function). It feels like those should be refactored into one.